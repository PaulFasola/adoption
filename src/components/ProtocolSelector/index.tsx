import React, { useEffect, useLayoutEffect, useMemo, useRef, useState } from 'react';
import { Button, ProtocolIcon, Symbol, DropDownList } from './style';
import { IProps, IProtocol } from './interfaces';
import { Icon, IconType } from '../common/Icon';
import { useLocale } from '../../hooks/useLocale';
import { IStrings } from './strings';

export const ProtocolSelector: React.FC<IProps> = (props) => {
  const dropdownRef = useRef<HTMLDivElement>(null);
  const strs = useLocale().strings.protocolSelector as IStrings;

  const [protocol, setProtocol] = useState<IProtocol>();
  const [dropdownIsOpen, setDropdownIsOpen] = useState<boolean>(false);

  useEffect(() => {
    const _handleOuterClick = (e: MouseEvent): void => {
      if (dropdownRef.current && !dropdownRef.current.contains(e.target as Node)) {
        dropdownIsOpen && setDropdownIsOpen(false);
      }
    };

    document.addEventListener('click', _handleOuterClick);
    return () => {
      document.removeEventListener('click', _handleOuterClick);
    };
  }, [dropdownIsOpen]);

  useLayoutEffect(() => {
    setProtocol(props.current);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.current]);

  // TODO: use a proper e2e to test accessibility
  /* istanbul ignore next */
  const handleKeypress = (protocol: IProtocol) => (e: React.KeyboardEvent) => {
    if (e.key === 'Enter') {
      handleProtocolPick(protocol)();
    }
  };

  const handleProtocolPick = (protocol: IProtocol) => (): void => {
    if (typeof props.onChange === 'function') {
      props.onChange(protocol);
    }

    setProtocol(protocol);
    setDropdownIsOpen(false);
  };

  const fullLabel = `${protocol?.name} - ${protocol?.symbol}`;

  const button = useMemo(() => {
    if (protocol) {
      return (
        <Button
          title={fullLabel}
          disabled={props.disabled}
          onClick={() => setDropdownIsOpen(!dropdownIsOpen)}
        >
          <ProtocolIcon src={protocol.logoURI} title={fullLabel} />
          <Symbol>{protocol.symbol}</Symbol>
          {props.list.length > 0 && (
            <Icon
              defaultTheme
              type={IconType.ArrowDown}
              style={{ width: '0.8em' }}
              disabled={props.disabled}
            />
          )}
        </Button>
      );
    }

    return (
      <Button disabled={props.disabled} onClick={() => setDropdownIsOpen(!dropdownIsOpen)}>
        <div>{strs.selectProtocol}</div>
        <Icon
          defaultTheme
          type={IconType.ArrowDown}
          style={{ width: '0.6em' }}
          disabled={props.disabled}
        />
      </Button>
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [protocol, props.disabled]);

  return (
    <>
      {button}
      <DropDownList data-test='dropdown' open={dropdownIsOpen} tabIndex={0} ref={dropdownRef}>
        {props.list.map((protocol, i) => (
          <div
            key={i}
            onClick={handleProtocolPick(protocol)}
            onKeyPress={handleKeypress(protocol)}
            tabIndex={0}
          >
            <ProtocolIcon src={protocol.logoURI} title={fullLabel} />
            <div>{protocol.name}</div>
            <div>{protocol.symbol}</div>
          </div>
        ))}
      </DropDownList>
    </>
  );
};
