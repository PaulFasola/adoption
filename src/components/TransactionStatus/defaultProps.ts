import { IProps } from './interfaces';
import { TxStatus } from './txStatus';

export const defaultColorMap: Record<TxStatus, string> = {
	completed: '#f08a5d',
	pending: 'deepskyblue',
	unknown: 'grey',
	failed: 'red'
}

export const defaultProps: Partial<IProps> = {
	date: {
		locale: 'en-US',
		options: {
			day: 'numeric',
			month: 'numeric',
			hour: 'numeric',
			minute: 'numeric'
		}
	}
}