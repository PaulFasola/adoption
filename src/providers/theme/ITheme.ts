export interface ITheme {
  primary: Partial<{
    backgroundColor: string;
    color: string;
    rule: string;
    boxShadow: string;
  }>;
}

export type PaletteCategory = 'primary' | 'secondary';
export type StyleProp = 'backgroundColor' | 'color' | 'boxShadow' | 'rule';
