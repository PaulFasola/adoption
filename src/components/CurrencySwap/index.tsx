import React, { useCallback, useEffect, useLayoutEffect, useState } from 'react';
import { Container, SwapButton, SwapInput, InputWrapper, SubmitButton, Overview } from './style';
import { IProps, IProtocolArrayPipe, IProtocolPipe, ProtocolEnd, ISwapValues } from './interfaces';
import { Icon, IconType } from '../common/Icon';
import { ProtocolSelector } from '../ProtocolSelector';
import { IProtocol } from '../ProtocolSelector/interfaces';
import { usePrevious } from 'react-delta';
import { preventCommonSymbol } from './utils';

const DEFAULT_MAX_FRACTION_DIGITS = 8;

const CurrencySwap: React.FC<IProps> = (props) => {
  const [protocols, setProtocols] = useState<IProtocolArrayPipe>({ input: [], output: [] });
  const [activeProtocols, setActiveProtocols] = useState<IProtocolPipe>();
  const [swapValues, setSwapValues] = useState<ISwapValues>({});
  const [canSwap, setCanSwap] = useState<boolean>(false);

  const dataRef = usePrevious(activeProtocols);

  /* istanbul ignore next */
  useLayoutEffect(() => {
    const { input, output } = props.protocols;
    const inputList = input ?? [];

    setProtocols({
      input: inputList,
      output: output ?? inputList,
    });
  }, [props.protocols]);

  const getMaxFractionDigits = useCallback(
    (decimals?: number): number => {
      const decimalsOrDefault = decimals ?? DEFAULT_MAX_FRACTION_DIGITS;

      if (!props.maxFractionDigits || props.maxFractionDigits >= decimalsOrDefault) {
        return decimalsOrDefault;
      }

      return props.maxFractionDigits;
    },
    [props.maxFractionDigits]
  );

  const reflectProtocolsAndPrices = useCallback(
    (end: ProtocolEnd, value: number): void => {
      if (!activeProtocols || !activeProtocols[end]) return;

      const currentProto = activeProtocols[end] as IProtocol;
      const oppositeEnd = end === 'input' ? 'output' : 'input';
      const oppositeProto = activeProtocols[oppositeEnd] as IProtocol | undefined;
      let oppositeValue = oppositeProto ? (value * currentProto.price) / oppositeProto.price : null;

      if (oppositeValue) {
        const toStr = oppositeValue.toLocaleString(undefined, {
          maximumFractionDigits: getMaxFractionDigits(activeProtocols[end]?.decimals),
        });
        oppositeValue = parseFloat(toStr.replace(/[^\d\.\-]/g, ''));
      }

      setSwapValues((prevProps) => ({
        ...prevProps,
        [end]: value,
        [oppositeEnd]: oppositeValue,
      }));
    },
    [activeProtocols, getMaxFractionDigits]
  );

  useEffect(() => {
    setProtocols((prevState) => {
      return {
        ...prevState,
        ...preventCommonSymbol(prevState, activeProtocols),
      };
    });

    const end: ProtocolEnd = dataRef?.input !== activeProtocols?.input ? 'output' : 'input';
    reflectProtocolsAndPrices(end, swapValues[end] ?? 0);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [activeProtocols, reflectProtocolsAndPrices]);

  /**
   * Determines if swapping is possible
   */
  useEffect((): void => {
    if (
      !activeProtocols ||
      !activeProtocols.input ||
      !activeProtocols.output ||
      !swapValues.input
    ) {
      return setCanSwap(false);
    }

    if (
      (activeProtocols.input.balance ?? -1) <= 0 ||
      (swapValues.input ?? -1) <= 0 ||
      swapValues.input > activeProtocols.input.balance
    ) {
      return setCanSwap(false);
    }

    setCanSwap(props.locked ?? true);
  }, [props.locked, activeProtocols, swapValues]);

  const handleValueChange = (end: ProtocolEnd) => (newValue: string): void => {
    const value = Number(newValue);
    setSwapValues({ input: undefined, output: undefined });

    if (isNaN(value)) return;
    reflectProtocolsAndPrices(end, value);
  };

  const handleProtocolChange = (end: ProtocolEnd) => (newProto: IProtocol) => {
    setActiveProtocols({ ...activeProtocols, ...{ [end]: newProto } });
  };

  const handleSubmit = () => {
    if (!swapValues || !activeProtocols) {
      return;
    }

    if (typeof props.onSubmit === 'function') {
      props.onSubmit();
    }

    setSwapValues({
      input: 0,
      output: 0,
    });
  };

  const getPriceEquiv = (): React.ReactNode => {
    if (!activeProtocols?.input || !activeProtocols?.output) {
      return;
    }

    const { input, output } = activeProtocols;
    if (input.price <= 0) {
      return;
    }

    return (
      <Overview>
        1 {input.symbol} =&nbsp;
        {(input.price / output.price).toLocaleString(undefined, {
          maximumFractionDigits: getMaxFractionDigits(output.decimals),
        })}
        &nbsp;
        {output.symbol}
      </Overview>
    );
  };

  const getButtonValue = (): string => {
    if (props.locked) {
      return 'Unlock Wallet';
    }

    if (
      swapValues.input &&
      activeProtocols?.input &&
      typeof activeProtocols.input.balance === 'number'
    ) {
      if (activeProtocols.input.balance < swapValues.input) {
        return 'Insufficient balance';
      }
    }

    return 'Swap!';
  };

  return (
    <Container noShadow={props.noShadow}>
      <InputWrapper>
        <div>
          <span>From</span>
          {activeProtocols?.input && (
            <Overview>Balance: {activeProtocols.input.balance ?? 0}</Overview>
          )}
        </div>
        <div>
          <SwapInput
            type='decimal'
            maxLength={19}
            placeholder='0.00'
            value={swapValues?.input}
            onValueChange={handleValueChange('input')}
          />
          <div>
            <ProtocolSelector
              current={activeProtocols?.input}
              list={protocols.input.filter((x) => !x.hidden)}
              onChange={handleProtocolChange('input')}
            />
          </div>
        </div>
      </InputWrapper>
      <SwapButton aria-label='Swap values'>
        <Icon
          type={IconType.ArrowDown}
          style={{ width: '18px' }}
          onClick={() => {
            setSwapValues({ input: swapValues.output, output: swapValues.input });
            setActiveProtocols({ input: activeProtocols?.output, output: activeProtocols?.input });
          }}
        />
      </SwapButton>
      <InputWrapper>
        <span>To</span>
        {getPriceEquiv()}
        <div>
          <SwapInput
            type='decimal'
            maxLength={19}
            placeholder='0.00'
            value={swapValues?.output}
            onValueChange={handleValueChange('output')}
          />
          <div>
            <ProtocolSelector
              current={activeProtocols?.output}
              list={protocols.output.filter((x) => !x.hidden)}
              onChange={handleProtocolChange('output')}
            />
          </div>
        </div>
      </InputWrapper>
      <SubmitButton aria-label='Submit' disabled={!canSwap} onClick={handleSubmit}>
        {getButtonValue()}
      </SubmitButton>
    </Container>
  );
};

CurrencySwap.defaultProps = {
  maxFractionDigits: DEFAULT_MAX_FRACTION_DIGITS,
};

export { CurrencySwap };
