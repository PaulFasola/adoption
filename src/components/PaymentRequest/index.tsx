import React, { Fragment, useLayoutEffect, useState } from 'react';
import { CancelButton, Container, DetailedView, Header, HelpLink, Request, Spinner, TransactionList, Visual, QRCode } from './style';
import { AnimatedIcon, IconType as AnimatedIconType } from '../common/AnimatedIcon';
import { defaultStrings, defaultProps } from './defaultProps';
import { AdaptiveSpan } from '../common/AdaptiveSpan';
import { PaymentStatus } from './enums/paymentStatus';
import { padDigits } from '../../utils/arithmetic';
import { shortenHash } from '../../utils/string';
import { IProps, IStrings } from './interfaces';
import { Icon, IconType } from '../common/Icon';
import { Item } from './item';

const PaymentRequest: React.FC<IProps> = (props) => {
	const remainingAmount = props.amount.toPay - (props.amount.received ?? 0);
	const [strings, setStrings] = useState<IStrings>(defaultStrings);

	useLayoutEffect(() => {
		/* istanbul ignore next */
		if (props.strings) {
			setStrings(s => { return { ...s, ...props.strings } });
		}
	}, [props.strings]);

	const _getFromDate = (date: Date): string => {
		const currentDate = new Date();
		const dayCount = (date.getTime() - currentDate.getTime()) / (1000 * 3600 * 24);

		let unit: Intl.RelativeTimeFormatUnit = 'day';
		let value = dayCount;

		if (dayCount >= 0 && dayCount <= 1) {
			const hoursCount = Math.abs(currentDate.getTime() - date.getTime()) / 36e5;
			unit = 'hours';
			value = hoursCount;

			if (hoursCount <= 1) {
				const minCount = Math.abs(currentDate.getTime() - date.getTime()) / 1000 / 60;
				unit = 'minute';
				value = Math.ceil(minCount);
			}
		}

		/* istanbul ignore next */
		return new Intl.RelativeTimeFormat(props.deadline?.dateLocale ?? 'en-US')
			.format(Math.ceil(value), unit)
	}

	const _getStatus = (): string => {
		if (remainingAmount <= 0) {
			return strings.txStatus.complete;
		} else if (props.customStatusText) {
			return props.customStatusText;
		}

		return strings.txStatus[props.status!];
	}

	const _renderVisual = (): React.ReactNode => {
		if (props.status && [PaymentStatus.FAILED, PaymentStatus.COMPLETE].includes(props.status)) {
			return <AnimatedIcon key={props.status}
				type={props.status === PaymentStatus.FAILED ? AnimatedIconType.Failure : AnimatedIconType.Sucess}
				style={{ display: 'flex', margin: 'auto' }}
			/>;
		} else if (props.showQRCode) {
			return (
				<QRCode
					bgColor="transparent"
					fgColor="#000000"
					level="L"
					style={{ width: 100 }}
					value={props.address} />
			)
		}
		return null;
	}

	return (
		<Container>
			<Header>
				<img src={props.logos.coin} alt={`${props.symbol} logo`} />
				{props.logos.company && <img src={props.logos.company} alt={`Company logo`} />}
			</Header>
			<div>
				<Request>
					<AdaptiveSpan
						text={strings.request}
						mapping={[
							{ tag: '{amount}', value: props.amount.toPay, style: 'bold' },
							{ tag: '{symbol}', value: props.symbol }
						]}
					/>
					<div title={props.address}>{props.address}</div>
				</Request>
				<Visual>
					{_renderVisual()}
				</Visual>
				<DetailedView>
					{props.sellerName ? <Item title={strings.seller}><span title={props.sellerName}>{props.sellerName}</span></Item> : null}
					{typeof props.amount.received === 'number' ? <Fragment>
						<Item title={strings.receivedAmount}>{padDigits(props.amount.received, props.decimalPlaces)} {props.symbol}</Item>
						<Item title={strings.remainingAmount}>{padDigits(remainingAmount, props.decimalPlaces)} {props.symbol}</Item>
					</Fragment> : null}
					{props.deadline && remainingAmount > 0 ? <Item title={strings.deadline}>
						{new Intl.DateTimeFormat(props.deadline.dateLocale ?? 'en-US').format(props.deadline.datetime)}
						{props.deadline.humanized && <span>&nbsp;({_getFromDate(props.deadline.datetime)})</span>}
					</Item> : null}
					{props.transactions?.length ? <Item title={strings.transactions.replace('{txAmount}', props.transactions.length.toString())}>
						<TransactionList>
							{props.transactions.map((transaction, index) => (
								<li key={index}>
									<a href={transaction.txUrl} title="Transaction" rel="noreferrer" target="_blank">
										{shortenHash(transaction.txHash, 3)}
									</a>
									<i>({padDigits(transaction.amount, props.decimalPlaces)} {props.symbol})</i>
								</li>
							))}
						</TransactionList>
					</Item> : null}
					<Item title={strings.status}>
						{(props.waitAnimation && props.status === PaymentStatus.PENDING) && <Spinner />}
						<AdaptiveSpan showTitle text={_getStatus()} style='bold' />
					</Item>
				</DetailedView>
			</div>
			<div>
				{props.onCancel && <CancelButton onClick={props.onCancel}>{strings.cancel}</CancelButton>}
				{props.helpUrl && <HelpLink href={props.helpUrl} title={strings.help} target="blank">
					<Icon type={IconType.HelpCircleO} style={{ width: 13, lineHeight: '15px' }} />{strings.help}
				</HelpLink>}
			</div>
		</Container>
	);
};

PaymentRequest.defaultProps = defaultProps;

export { PaymentRequest };