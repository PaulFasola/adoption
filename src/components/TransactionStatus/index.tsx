import React, { useLayoutEffect, useState } from 'react';
import {
  Amount,
  Arrow,
  Container,
  CustomComponent,
  Date,
  DetailedView,
  IconWrapper,
  Row,
  Side,
  Status,
  StatusBar,
  Trajectory,
  TxFees,
} from './style';
import { defaultColorMap, defaultProps } from './defaultProps';
import { AdaptiveSpan } from '../common/AdaptiveSpan';
import { shortenHash } from '../../utils/string';
import { Icon, IconType } from '../common/Icon';
import { IAdress, IProps } from './interfaces';
import { TxStatus } from './txStatus';

const TransactionStatus: React.FC<IProps> = (props) => {
  const [status, setStatus] = useState<TxStatus>(TxStatus.UNKNOWN);
  const [colorMap, setColorMap] = useState<Record<TxStatus, string>>(defaultColorMap);
  const [isToggled, setIsToggled] = useState<boolean>(false);

  const shouldRenderDetails = Boolean(
    props.txFees || props.sender || props.receiver || props.customDetailComponent
  );
  const shouldRenderTrajectory = Boolean(props.sender || props.receiver);
  const shouldRenderSide = Boolean(props.txURL || shouldRenderDetails);

  /* istanbul ignore next */
  useLayoutEffect(() => {
    if (props.status) {
      setStatus(props.status);
    }

    if (props.autoShowDetails && shouldRenderDetails) {
      setIsToggled(true);
    }

    setColorMap({ ...defaultColorMap, ...props.colorMap });
  }, [props.status, props.colorMap, props.autoShowDetails, shouldRenderDetails]);

  const _handleToggleClick = (): void => setIsToggled(!isToggled);

  const _getStatus = (): string => {
    let output = status.toUpperCase() + status.slice(1);
    if (!props.uncapitalizeStatus) output = status.toUpperCase();

    return output;
  };

  const _getTransaction = (tx?: IAdress): React.ReactNode => {
    if (!tx || !tx.hash) return null;

    return (
      <a href={tx.url} rel='noreferrer' target='_blank'>
        {shortenHash(tx.hash, 7)}
      </a>
    );
  };

  return (
    <Container showDetails={isToggled} hasDate={Boolean(props.date?.value)}>
      <StatusBar
        backgroundColor={colorMap[status]}
        animate={props.animated ? props.status : null}
      />
      <Row showSide={shouldRenderSide}>
        <Amount>
          <AdaptiveSpan
            text='<h1>{amount}</h1><p>{symbol}</p>'
            mapping={[
              { tag: '{amount}', value: props.amount, style: 'bold' },
              { tag: '{symbol}', value: props.symbol },
            ]}
          />
        </Amount>
        <Status title={`Status`}>{_getStatus()}</Status>
        {shouldRenderSide && (
          <Side>
            {props.txURL && (
              <IconWrapper>
                <Icon
                  type={IconType.OutboundLink}
                  style={{ width: 21 }}
                  url={props.txURL}
                  targetBlank
                />
              </IconWrapper>
            )}
            {shouldRenderDetails && (
              <IconWrapper clickable onClick={_handleToggleClick}>
                <Icon
                  type={isToggled ? IconType.ArrowUp : IconType.ArrowDown}
                  style={{ width: 15 }}
                />
              </IconWrapper>
            )}
          </Side>
        )}
      </Row>
      {shouldRenderDetails && (
        <DetailedView>
          {shouldRenderTrajectory && (
            <div>
              <Trajectory>
                {_getTransaction(props.sender)}
                {props.sender && props.receiver && <Arrow />}
                {_getTransaction(props.receiver)}
              </Trajectory>
              <TxFees>
                {props.txFees} {props.symbol} (fees)
              </TxFees>
            </div>
          )}
          <CustomComponent>{props.customDetailComponent}</CustomComponent>
        </DetailedView>
      )}
      {props.date && props.date.value && (
        <Date>
          <span>
            {new Intl.DateTimeFormat(props.date.locale ?? 'en-US', props.date.options).format(
              props.date.value
            )}
          </span>
        </Date>
      )}
    </Container>
  );
};

TransactionStatus.defaultProps = defaultProps;

export { TransactionStatus, TxStatus };
