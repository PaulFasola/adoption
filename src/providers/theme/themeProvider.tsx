import React, { useCallback, useEffect, useState } from 'react';
import { ThemeProvider as SCThemeProvider } from 'styled-components';
import { defaultThemes, ThemeLabel } from './defaultThemes';
import { IViableTheme, IThemeContext } from './IThemeContext';
import { mergeDeep } from '../../utils/mergeDeep';
import { ITheme } from './ITheme';

const LOCAL_STORAGE_KEY = 'adoption-theme';

interface IProps {
  customThemes?: Record<string, Partial<ITheme>>;
}

export const defaultTheme = {
  name: 'light',
  palette: defaultThemes.light,
};

export const ThemeContext = React.createContext<IThemeContext>({
  currentTheme: defaultTheme,
  switchTo: () => { },
});

export const ThemeProvider: React.FC<IProps> = ({ customThemes, children }) => {
  const [availableThemes, setAvailableThemes] = useState<Record<string, ITheme>>(defaultThemes);
  const [currentTheme, setCurrentTheme] = useState<IViableTheme>(defaultTheme);

  const setViableThemeOrDefault = useCallback(
    (themeName: string): [string, ITheme] => {
      const requestedTheme = availableThemes[themeName] as ITheme;
      let theme = defaultTheme;

      if (requestedTheme) {
        theme = {
          name: themeName,
          palette: requestedTheme,
        };
      } else {
        console.warn(`[WARN] Adoption Theme - requested theme "${themeName}" was not found. Defaulting to "${theme.name
          }" preset.\n
Add your theme to 'customThemes' property on <ThemeProvider>.
Available themes: ${Object.keys(availableThemes).join(', ')}`);
      }

      setCurrentTheme(theme);

      return [theme.name, theme.palette];
    },
    [availableThemes]
  );

  useEffect(() => {
    const theme = localStorage.getItem(LOCAL_STORAGE_KEY) ?? 'light';
    setViableThemeOrDefault(theme);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (!customThemes || Object.keys(customThemes).length === 0) return;
    const filledCustomThemes: Record<string, ITheme> = {};

    Object.keys(customThemes).forEach((key) => {
      const viableTheme = (mergeDeep(defaultTheme.palette, customThemes[key]) as unknown) as ITheme;
      filledCustomThemes[key] = viableTheme;
    });

    setAvailableThemes((prevProps) => {
      return { ...prevProps, ...filledCustomThemes };
    });
  }, [customThemes]);

  const switchTo = (nextTheme: ThemeLabel | string) => {
    const [name] = setViableThemeOrDefault(nextTheme);
    localStorage.setItem(LOCAL_STORAGE_KEY, name);
  };

  return (
    <ThemeContext.Provider value={{ currentTheme, switchTo }}>
      <SCThemeProvider theme={currentTheme.palette}>{children}</SCThemeProvider>
    </ThemeContext.Provider>
  );
};
