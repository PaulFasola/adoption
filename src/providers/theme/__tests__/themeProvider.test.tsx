import React, { useContext } from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import { configure, shallow } from 'enzyme';
import { ThemeContext, ThemeProvider } from '../themeProvider';
import { TransactionStatus, TxStatus } from '../../../components/TransactionStatus';
import { ITheme } from '../ITheme';

expect.extend(toHaveNoViolations);
configure({ adapter: new Adapter() });

describe('ThemeProvider component', () => {
  it('should render', () => {
    shallow(
      <ThemeProvider>
        <TransactionStatus amount='0.0001' symbol='BTC' status={TxStatus.PENDING} />
      </ThemeProvider>
    );
  });

  it('should fallback to the default theme if the requested one is missing', async () => {
    const MISSING_THEME_LABEL = 'missingTheme';

    const MyComponent = (): React.ReactElement => {
      const { currentTheme, switchTo } = useContext(ThemeContext);

      return (
        <>
          {currentTheme.name}
          <button data-testid='switch-theme' onClick={() => switchTo(MISSING_THEME_LABEL)}>
            switch
          </button>
          <TransactionStatus amount='0.0001' symbol='BTC' status={TxStatus.PENDING} />
        </>
      );
    };

    const component = render(
      <ThemeProvider>
        <MyComponent />
      </ThemeProvider>
    );

    const consoleOutput: string[] = [];
    const mockedWarn = (output: string) => consoleOutput.push(output);

    const button = await component.findByTestId('switch-theme');
    const originalWarn = console.warn;

    console.warn = mockedWarn;
    button.click();
    console.warn = originalWarn;

    expect(consoleOutput.length).toEqual(1);
    expect(consoleOutput[0].indexOf(MISSING_THEME_LABEL)).toBeGreaterThan(-1);
  });

  it('should render custom themes when requested', async () => {
    const appTheme: Record<string, Partial<ITheme>> = {
      myTheme: {
        primary: {
          backgroundColor: 'blue',
        },
      },
    };

    const MyComponent = (): React.ReactElement => {
      const { currentTheme, switchTo } = useContext(ThemeContext);

      return (
        <>
          {currentTheme.name}
          <button data-testid='switch-theme' onClick={() => switchTo('shouldFail')}>
            switch
          </button>
          <TransactionStatus amount='0.0001' symbol='BTC' status={TxStatus.PENDING} />
        </>
      );
    };

    const component = render(
      <ThemeProvider customThemes={appTheme}>
        <MyComponent />
      </ThemeProvider>
    );
    const button = await component.findByTestId('switch-theme');
    button.click();

    expect(component).toMatchSnapshot();
  });

  it('should not have any accessibility issues', async () => {
    const { container } = render(
      <ThemeProvider>
        <TransactionStatus amount='0.0001' symbol='BTC' status={TxStatus.PENDING} />
      </ThemeProvider>
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});
