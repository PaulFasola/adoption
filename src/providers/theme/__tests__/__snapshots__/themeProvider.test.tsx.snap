// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider component should render custom themes when requested 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      light
      <button
        data-testid="switch-theme"
      >
        switch
      </button>
      <div
        class="sc-bdfBwQ fgFrNy"
      >
        <div
          class="sc-iBPRYJ fHvgvO"
        />
        <div
          class="sc-dlfnbm bYjgCU"
        >
          <div
            class="sc-hKgILt LWOpd"
          >
            <span>
              <h1>
                <b>
                  0.0001
                </b>
              </h1>
              <p>
                BTC
              </p>
            </span>
          </div>
          <div
            class="sc-jSgupP djuvgz"
            title="Status"
          >
            PENDING
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    light
    <button
      data-testid="switch-theme"
    >
      switch
    </button>
    <div
      class="sc-bdfBwQ fgFrNy"
    >
      <div
        class="sc-iBPRYJ fHvgvO"
      />
      <div
        class="sc-dlfnbm bYjgCU"
      >
        <div
          class="sc-hKgILt LWOpd"
        >
          <span>
            <h1>
              <b>
                0.0001
              </b>
            </h1>
            <p>
              BTC
            </p>
          </span>
        </div>
        <div
          class="sc-jSgupP djuvgz"
          title="Status"
        >
          PENDING
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
